src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,540 @@
+            print(
+                "Early stopping based on learning rate scheduler callback (min_lr was reached)."
+            )
+            return False
+
+        pre_step_learning_rate = trainer.learning_rate
+        trainer.optimizer.set_learning_rate(
+            self.lr_scheduler(trainer.optimizer.num_update)
+        )
+
+        if not trainer.learning_rate == pre_step_learning_rate:
+            if best_epoch_info["epoch_no"] == -1:
+                raise GluonTSUserError(
+                    "Got NaN in first epoch. Try reducing initial learning rate."
+                )
+
+            logger.info(
+                f"Loading parameters from best epoch "
+                f"({best_epoch_info['epoch_no']})"
+            )
+            training_network.load_parameters(
+                best_epoch_info["params_path"], ctx
+            )
+
+        return True
+
+
+class ModelIterationAveraging(Callback):
+    """
+    Callback to implement iteration based model averaging strategies.
+
+    Parameters
+    ----------
+    avg_strategy
+        IterationAveragingStrategy, one of NTA or Alpha_Suffix from
+        gluonts.mx.trainer.model_iteration_averaging
+    """
+
+    @validated()
+    def __init__(self, avg_strategy: IterationAveragingStrategy):
+        self.avg_strategy = avg_strategy
+
+    def on_validation_epoch_start(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        # use averaged model for validation
+        self.avg_strategy.load_averaged_model(training_network)
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        self.avg_strategy.load_cached_model(training_network)
+        return True
+
+    def on_train_batch_end(self, training_network: nn.HybridBlock) -> None:
+        self.avg_strategy.apply(training_network)
+
+    def on_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+        best_epoch_info: Dict[str, Any],
+        ctx: mx.Context,
+    ) -> bool:
+        self.avg_strategy.update_average_trigger(
+            metric=epoch_loss, epoch=epoch_no + 1
+        )
+        # once triggered, update the average immediately
+        self.avg_strategy.apply(training_network)
+        return True
+
+    def on_train_end(
+        self,
+        training_network: nn.HybridBlock,
+        temporary_dir: str,
+        ctx: mx.context.Context = None,
+    ) -> None:
+        logging.info("Loading averaged parameters.")
+        self.avg_strategy.load_averaged_model(training_network)
+
+
+class ModelAveraging(Callback):
+    """
+    Callback to implement model averaging strategies.
+    Selects the checkpoints with the best loss values and computes the model average or weighted model average depending on the chosen avg_strategy.
```

Comment:
Can you make sure the docstring is within the 79 limit?

Reply:
Will do, but that is a drop in the ocean, it seems like that limit isn't being honored almost everywhere^^

Reply:
Not sure that's a convincing argument ;)

Reply:
Sure, I fixed it in the entire `gluonts.mx.trainer` module now ;)